lib/awful/screenshot.lua

```diff
@@ -0,0 +1,870 @@
+  if ss.screen then
+    if type(ss.screen) == "number" then
+      ss._private.screen = capi.screen[ss.screen] or aw_screen.focused()
+    end
+  else
+    ss._private.screen = aw_screen.focused()
+  end
+
+  ss._private.geometry = ss.screen.geometry
+  ss:filepath_builder()
+  ss._private.surface = gears.surface(ss.screen.content) -- surface has no setter
+
+  return ss
+
+end
+
+-- Internal function executed when a client window screenshot is taken.
+function screenshot_methods.client(ss)
+	--
+  -- note the use of _private becuase client has no setter
+  if not ss.client then
+    ss._private.client = capi.client.focus 
+  end
+
+  ss._private.geometry = ss.client:geometry()
+  ss:filepath_builder()
+  ss._private.surface = gears.surface(ss.client.content) -- surface has no setter
+  return ss
+end
+
+-- Internal function executed when a snipper screenshot tool is launched.
+function screenshot_methods.snipper(ss)
+
+  if type(ss._private.on_success_cb) ~= "function" then
+    ss._private.on_success_cb = default_on_success_cb -- the cb has no setter
+  end
+
+  local mg_callback = mk_mg_callback(ss)
+  capi.mousegrabber.run(mg_callback, "crosshair")
+
+  return true
+
+end
+
+-- Internal function executed when a snip screenshow (a defined geometry) is
+-- taken.
+function screenshot_methods.snip(ss)
+
+  local root_w, root_h
+  local root_intrsct
+  local snip_surf
+
+  root_w, root_h = root.size()
+
+  if not(ss.geometry and
+         type(ss.geometry.x) == "number" and
+         type(ss.geometry.y) == "number" and
+         type(ss.geometry.width) == "number" and
+         type(ss.geometry.height) == "number") then
+
+    -- Default to entire root window. Also geometry has no setter.
+    ss._private.geometry = {x = 0, y = 0, width = root_w, height = root_h}
+
+  end
+
+  root_intrsct = gears.geometry.rectangle.get_intersection(ss.geometry,
+                                                           {x = 0, y = 0,
+                                                            width = root_w,
+                                                            height = root_h})
+
+  snip_surf = crop_shot(root_intrsct.x, root_intrsct.y,
+                          root_intrsct.width, root_intrsct.height)
+
+  ss:filepath_builder()
+  ss._private.surface = gears.surface(snip_surf) -- surface has no setter
+
+  return ss
+
+end
+
+-- Default method is root
+screenshot_methods.default = screenshot_methods.root
+local default_method_name = "root"
+
+-- Module routines
+
+--- Function to initialize the screenshot library.
+--
+-- Currently only sets the screenshot directory, the filename, prefix and the
+-- snipper tool outline color. More initialization can be added as the API
+-- expands.
+-- @staticfct awful.screenshot.set_defaults
+-- @tparam[opt] table args Table passed with the configuration data.
+-- @treturn boolean true or false depending on success
+function module.set_defaults(args)
+
+  local tmp
```

Comment:
Please use `local tmp = check_directory(args.directory)` rather than 3 lines to improve readability.

Reply:
I will make this change if we keep this function, per above.